### PR TITLE
8307351: (CmpI/L(AndI/L reg1 reg2)) on x86 can be optimized

### PR DESCRIPTION
**Pull request by ichttt**
This patch aims to optimize a case where a And-Node followed by a Cmp-Node would not be converted into a single "test" instruction. The Architecture Description file currently only handles the cases where the right operand of the And-Node is a constant, but not if both are a register.
Before this patch, a "and" followed by a "test" would be emitted, so the removed "and" means 2 bytes less have to be emitted.
I've attached a JMH Benchmark to demonstrate the performance improvements. Here are the numbers of my Windows 11 machine:
Before:
```
Benchmark                                                   Mode  Cnt   Score   Error  Units
AndCmpTestInstruction.benchmarkOpaqueAndCmpEqualsInt        avgt    8  26,736 ± 0,131  ns/op
AndCmpTestInstruction.benchmarkOpaqueAndCmpEqualsLong       avgt    8  24,305 ± 0,610  ns/op
AndCmpTestInstruction.benchmarkStaticAndCmpEqualsInt        avgt    8  33,052 ± 0,056  ns/op
AndCmpTestInstruction.benchmarkStaticLargeAndCmpEqualsLong  avgt    8  18,355 ± 0,030  ns/op
AndCmpTestInstruction.benchmarkStaticSmallAndCmpEqualsLong  avgt    8  25,587 ± 0,107  ns/op
```
After:
```
Benchmark                                                   Mode  Cnt   Score   Error  Units  Improvement
AndCmpTestInstruction.benchmarkOpaqueAndCmpEqualsInt        avgt    8  22,665 ± 0,170  ns/op  (~18%)
AndCmpTestInstruction.benchmarkOpaqueAndCmpEqualsLong       avgt    8  18,880 ± 0,123  ns/op  (~29%)
AndCmpTestInstruction.benchmarkStaticAndCmpEqualsInt        avgt    8  33,198 ± 0,126  ns/op  (unchanged)
AndCmpTestInstruction.benchmarkStaticLargeAndCmpEqualsLong  avgt    8  18,427 ± 0,079  ns/op  (unchanged)
AndCmpTestInstruction.benchmarkStaticSmallAndCmpEqualsLong  avgt    8  25,641 ± 0,168  ns/op  (unchanged)
```
As you can see, the cases with a small static mask have not improved as they have already been covered by another match rule. The test with the large static mask (benchmarkStaticLargeAndCmpEqualsLong) has a sorted instruction sequence as the value is moved to a register and it is not being used directly in the instruction, but there is no measurable performance uplift here.
I've tested my changes using the Tier1 jtreg Tests on Windows.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8307351](https://bugs.openjdk.org/browse/JDK-8307351): (CmpI/L(AndI/L reg1 reg2)) on x86 can be optimized

### Reviewers
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**)
 * [Quan Anh Mai](https://openjdk.org/census#qamai) (@merykitty - Committer)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13587/head:pull/13587` \
`$ git checkout pull/13587`

Update a local copy of the PR: \
`$ git checkout pull/13587` \
`$ git pull https://git.openjdk.org/jdk.git pull/13587/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13587`

View PR using the GUI difftool: \
`$ git pr show -t 13587`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13587.diff">https://git.openjdk.org/jdk/pull/13587.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13587#issuecomment-1533193298)

### CODE REVIEW

**Comment by merykitty** [comment]
You can add other `Cmp` patterns: `cmp_mem_imm`, `test_mem` (which is basically `cmp_mem_imm` but matching `test_reg` helps cisc spilling), `test_mem_imm`. `cmp` patterns may need unsigned versions, too while `test` patterns do not since `CmpU x 0` should be idealised into `CmpI x 0` already.

Thanks.

**Comment by TobiHartmann** [comment]
A bug report was created that you can link to this PR by updating the title accordingly:
https://bugs.openjdk.org/browse/JDK-8307351

File: src/hotspot/cpu/x86/x86_64.ad [hunk 1]
```diff
@@ -12449,6 +12449,17 @@ instruct testI_reg_imm(rFlagsReg cr, rRegI src, immI con, immI_0 zero)
   ins_pipe(ialu_cr_reg_imm);
 %}
 
+instruct testI_reg_reg(rFlagsReg cr, rRegI src1, rRegI src2, immI_0 zero)
+%{
+  match(Set cr (CmpI (AndI src1 src2) zero));
+
+  format %{ "testl   $src1, $src2\t# long" %}
```
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86_64.ad:12456]: The format string has "# long" should be "# int" here as this is integer operation.

File: test/micro/org/openjdk/bench/vm/compiler/x86/AndCmpTestInstruction.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
```
Comment (sviswa7) [inline test/micro/org/openjdk/bench/vm/compiler/x86/AndCmpTestInstruction.java:2]: Copyright year should be 2023.

**Comment by sviswa7** [comment]
@ichttt Nice performance gain for opaque operations.

**Comment by ichttt** [comment]
Regarding operations on memory: I am not sure if this would be benificial, as (if I understand it correctly) these operations cannot be marco-fused. C++ compilers like clang or gcc also don't seem to emit test instructions with memory operands and immediates. So I would like to omit that from the pull request and limit it to register case, which never result in performance degradation.

**Review by sviswa7 — APPROVED** [review]
Looks good to me.

**Review by merykitty — APPROVED** [review]

**Review by TobiHartmann — APPROVED** [review]
Looks good to me too. Testing passed.

**Pull request closed without merging**